bindings/python/crocoddyl/utils/__init__.py

```diff
@@ -437,6 +437,167 @@ def __init__(self, model):
         self.differential = model.differential.createData(self)
 
 
+class IntegratedActionModelRK4Derived(crocoddyl.ActionModelAbstract):
+    def __init__(self, diffModel, timeStep=1e-3, withCostResiduals=True):
+        crocoddyl.ActionModelAbstract.__init__(self, diffModel.state, diffModel.nu, diffModel.nr)
+        self.differential = diffModel
+        self.timeStep = timeStep
+        self.rk4_inc = [0.5, 0.5, 1.]
+        self.nx = self.differential.state.nx
+        self.ndx = self.differential.state.ndx
+        self.nq = self.differential.state.nq
+        self.nv = self.differential.state.nv
+        self.enable_integration = (self.timeStep > 0.)
+
+    def createData(self):
+        return IntegratedActionDataRK4Derived(self)
+
+    def calc(self, data, x, u=None):
+        nq, dt = self.nq, self.timeStep
+
+        data.y[0] = x
+        for i in range(3):
+            self.differential.calc(data.differential[i], data.y[i], u)
+            data.acc[i] = data.differential[i].xout
+            data.int[i] = data.differential[i].cost
+            data.ki[i] = np.concatenate([data.y[i][nq:], data.acc[i]])
+            data.y[i + 1] = self.differential.state.integrate(x, data.ki[i] * self.rk4_inc[i] * dt)
+
+        self.differential.calc(data.differential[3], data.y[3], u)
+        data.acc[3] = data.differential[3].xout
+        data.int[3] = data.differential[3].cost
+        data.ki[3] = np.concatenate([data.y[3][nq:], data.acc[3]])
+        if (self.enable_integration):
+            data.dx = (data.ki[0] + 2. * data.ki[1] + 2. * data.ki[2] + data.ki[3]) * dt / 6
+            data.xnext = self.differential.state.integrate(x, data.dx)
+            data.cost = (data.int[0] + 2 * data.int[1] + 2 * data.int[2] + data.int[3]) * dt / 6
+        else:
+            data.dx = np.zeros([self.ndx])
+            data.xnext = x
+            data.cost = data.differential[0].cost
+
+        return data.xnext, data.cost
+
+    def calcDiff(self, data, x, u):
+        ndx, nu, nv, dt = self.ndx, self.nu, self.nv, self.timeStep
+        for i in range(4):
+            self.differential.calcDiff(data.differential[i], data.y[i], u)
+            data.dki_dy[i] = np.bmat([[np.zeros([nv, nv]), np.identity(nv)], [data.differential[i].Fx]])
+
+        data.dki_du[0] = np.vstack([np.zeros([nv, nu]), data.differential[0].Fu])
+
+        data.Lx[:] = data.differential[0].Lx
+        data.Lu[:] = data.differential[0].Lu
+
+        data.dy_dx[0] = np.identity(nv * 2)
```

Comment:
This is tiny detail about the use of the Python bindings. If you do
```cpp
data.dy_dx[0] = np.identity(nv * 2)
```
what it is occurring is that Python is allocating a new object, instead of copying internally the data.

To solve this, you should write the internal data, e.g.
```cpp
data.dy_dx[0][:, :] = np.identity(nv * 2)
```
which it is important to run in parallel this code (but this feature is not yet available).
Feel free to make these changes, however, do not worry because it is not that important.

Btw, the previous operation is needed to be performed once, right? If that is correct then it makes sense to do it in the data creation. With this approach, you avoid to fill constant values. If you have this in the c++ code, then I will suggest to do it there to improve efficiency (not needed to do it in this Python class).

Reply:
done